### PR TITLE
Leverage generics to make the `ListResources` api more user friendly

### PR DESCRIPTION
**Pull request by rosstimothy**
Add `client.GetAllResources[T types.ResourceWithLabels]` and `client.GetResourcePage[T types.ResourceWithLabels]` to make getting resources less verbose:

#### Current API
```go
	resources, err := GetResourcesWithFilters(ctx, c, proto.ListResourcesRequest{
		ResourceType: types.KindNode,
		Namespace:    namespace,
	})	
	if err != nil {
		return nil, trace.Wrap(err)
	}
	servers, err := types.ResourcesWithLabels(resources).AsServers()

	return servers, nil

```

#### API with the changes in this PR

```go
	servers, err := GetAllResources[types.Server](ctx, c, &proto.ListResourcesRequest{
		ResourceType: types.KindNode,
		Namespace:    namespace,
	})
	return servers, trace.Wrap(err)

### CODE REVIEW

File: api/client/client_test.go [hunk 1]
```diff
@@ -224,10 +225,12 @@ func testResources(resourceType, namespace string) ([]types.ResourceWithLabels,
 				Hostname: "localhost",
 				HostID:   fmt.Sprintf("host-%d", i),
 			})
-
 			if err != nil {
 				return nil, trace.Wrap(err)
 			}
+
+			resourceWithLabels = resource
+			resources = append(resources, resourceWithLabels.(T))
```
Comment (greedy52) [inline api/client/client_test.go:233]: probably can also do `resources = append(resources, any(resource).(T))`

File: api/client/client_test.go [hunk 2]
```diff
@@ -577,6 +591,186 @@ func TestGetResources(t *testing.T) {
 	clt, err := srv.NewClient(ctx)
 	require.NoError(t, err)
 
+	t.Run("DatabaseServer", func(t *testing.T) {
```
Comment (greedy52) [inline api/client/client_test.go:594]: nit: any differences between these tests? maybe can have a generic helper function to avoid repeating.

**Review by greedy52 — APPROVED** [review]
are we deprecating the non-generics version eventually?

File: api/client/client.go [hunk 3]
```diff
@@ -2783,6 +2748,108 @@ func (c *Client) ListResources(ctx context.Context, req proto.ListResourcesReque
...
+				return nil, 0, trace.BadParameter("received unexpected resource type %T", resource)
+			}
+			out = append(out, t)
+		}
+
+		return out, int(resp.TotalCount), nil
+	}
+}
+
+// GetAllResources is a helper for getting all existing resources that match the provided request.In addition to
```
Comment (ravicious) [inline api/client/client.go:2830]: Missing space.

```suggestion
// GetAllResources is a helper for getting all existing resources that match the provided request. In addition to
```

File: lib/teleterm/clusters/cluster_servers.go [hunk 4]
```diff
@@ -59,18 +71,8 @@ func (c *Cluster) GetServers(ctx context.Context, r *api.GetServersRequest) (*Ge
...
-			ResourceType:        types.KindNode,
-			Limit:               r.Limit,
-			SortBy:              sortBy,
-			StartKey:            r.StartKey,
-			PredicateExpression: r.Query,
-			SearchKeywords:      client.ParseSearchKeywords(r.Search, ' '),
-			UseSearchAsRoles:    r.SearchAsRoles == "yes",
-		})
+
+		servers, err = apiclient.GetAllResources[types.Server](ctx, authClient, req)
```
Comment (ravicious) [inline lib/teleterm/clusters/cluster_servers.go:75]: I'm pretty sure `Cluster.GetServers` should not use `apiclient.GetAllResources`. This method is used to display the paginated table in the cluster view in Connect, so we want to fetch only one table at a time.

I was about to ask if there's a specific reason behind changing just `Cluster.GetServers` but not `Cluster.GetDatabases` or `Cluster.GetKubes`, but it seems like all three places should continue using `authClient.ListResources` (or be refactored to use `apiclient.GetResourcePage`) as they're all responsible for feeding data for those resource tables.

File: lib/web/apps.go [hunk 5]
```diff
@@ -51,12 +52,12 @@ func (h *Handler) clusterAppsGet(w http.ResponseWriter, r *http.Request, p httpr
...
 	}
 
-	resp, err := listResources(clt, r, types.KindAppServer)
+	req, err := convertListResourcesRequest(r, types.KindAppServer)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	appServers, err := types.ResourcesWithLabels(resp.Resources).AsAppServers()
+	appServers, total, err := client.GetResourcePage[types.AppServer](r.Context(), clt, req)
```
Comment (ravicious) [inline lib/web/apps.go:60]: It looks like the implementation of `client.GetResourcePage` is significantly different than that of `ListResources`. `GetResourcePage` ignores the limit that is passed with `req` and returns all matching resources.

File: lib/web/apps.go [hunk 6]
```diff
@@ -74,8 +75,8 @@ func (h *Handler) clusterAppsGet(w http.ResponseWriter, r *http.Request, p httpr
 			Identity:          identity,
 			Apps:              apps,
 		}),
-		StartKey:   resp.NextKey,
-		TotalCount: resp.TotalCount,
+		StartKey:   req.StartKey,
```
Comment (ravicious) [inline lib/web/apps.go:78]: I think in the Web UI the `StartKey` included in the response is the key that the client is supposed to use as the next `StartKey` if they want to fetch another page. It's a little bit confusing but for now we have to preserve this behavior.

File: tool/tsh/app.go [hunk 7]
```diff
@@ -266,6 +266,7 @@ func getRegisteredApp(cf *CLIConf, tc *client.TeleportClient) (app types.Applica
 	err = client.RetryWithRelogin(cf.Context, tc, func() error {
 		apps, err = tc.ListApps(cf.Context, &proto.ListResourcesRequest{
 			Namespace:           tc.Namespace,
+			ResourceType:        types.KindAppServer,
```
Comment (ravicious) [inline tool/tsh/app.go:269]: Is there a reason why methods like `tc.ListApps` should no longer set `ResourceType` by themselves and instead expect the caller to set `ResourceType` on the request?

**Review by ravicious — COMMENTED** [review]
The changes in lib/teleterm break listing servers and I'm pretty sure they break listing some resources in the Web UI as well. I'm going to ask folks on #dev-frontend more familiar with resource listing and pagination in the Web UI to have another look at it.

**Review by kimlisa — COMMENTED** [review]
lgtm, but i'd wait for rafal to finish reviewing (i think today is a holiday for him)

**Review by ravicious — APPROVED** [review]
Left one suggestion and there seems to be a minor issue where the wrong resource type is used but other than that it looks good. Thanks for taking care of refactoring all of those call sites.

**Pull request merged**